// tsparticles/umd/Interactions/External/Attract/Attractor.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../../../Utils/NumberUtils", "../../../Core/Utils/Circle", "../../../Core/Utils/Constants", "../../../Core/Utils/ExternalInteractorBase", "../../../Core/Utils/Vector", "../../../Utils/Utils"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.Attractor = void 0;
    const NumberUtils_1 = require("../../../Utils/NumberUtils");
    const Circle_1 = require("../../../Core/Utils/Circle");
    const Constants_1 = require("../../../Core/Utils/Constants");
    const ExternalInteractorBase_1 = require("../../../Core/Utils/ExternalInteractorBase");
    const Vector_1 = require("../../../Core/Utils/Vector");
    const Utils_1 = require("../../../Utils/Utils");
    class Attractor extends ExternalInteractorBase_1.ExternalInteractorBase {
        constructor(container) {
            super(container);
            if (!container.attract) {
                container.attract = { particles: [] };
            }
            this.handleClickMode = (mode) => {
                const options = this.container.actualOptions;
                if (mode !== "attract") {
                    return;
                }
                if (!container.attract) {
                    container.attract = { particles: [] };
                }
                container.attract.clicking = true;
                container.attract.count = 0;
                for (const particle of container.attract.particles) {
                    particle.velocity.setTo(particle.initialVelocity);
                }
                container.attract.particles = [];
                container.attract.finish = false;
                setTimeout(() => {
                    if (!container.destroyed) {
                        if (!container.attract) {
                            container.attract = { particles: [] };
                        }
                        container.attract.clicking = false;
                    }
                }, options.interactivity.modes.attract.duration * 1000);
            };
        }
        isEnabled() {
            const container = this.container, options = container.actualOptions, mouse = container.interactivity.mouse, events = options.interactivity.events;
            if ((!mouse.position || !events.onHover.enable) && (!mouse.clickPosition || !events.onClick.enable)) {
                return false;
            }
            const hoverMode = events.onHover.mode, clickMode = events.onClick.mode;
            return (0, Utils_1.isInArray)("attract", hoverMode) || (0, Utils_1.isInArray)("attract", clickMode);
        }
        reset() {
        }
        async interact() {
            const container = this.container, options = container.actualOptions, mouseMoveStatus = container.interactivity.status === Constants_1.Constants.mouseMoveEvent, events = options.interactivity.events, hoverEnabled = events.onHover.enable, hoverMode = events.onHover.mode, clickEnabled = events.onClick.enable, clickMode = events.onClick.mode;
            if (mouseMoveStatus && hoverEnabled && (0, Utils_1.isInArray)("attract", hoverMode)) {
                this.hoverAttract();
            }
            else if (clickEnabled && (0, Utils_1.isInArray)("attract", clickMode)) {
                this.clickAttract();
            }
        }
        hoverAttract() {
            const container = this.container;
            const mousePos = container.interactivity.mouse.position;
            if (!mousePos) {
                return;
            }
            const attractRadius = container.retina.attractModeDistance;
            this.processAttract(mousePos, attractRadius, new Circle_1.Circle(mousePos.x, mousePos.y, attractRadius));
        }
        processAttract(position, attractRadius, area) {
            const container = this.container;
            const attractOptions = container.actualOptions.interactivity.modes.attract;
            const query = container.particles.quadTree.query(area);
            for (const particle of query) {
                const { dx, dy, distance } = (0, NumberUtils_1.getDistances)(particle.position, position);
                const velocity = attractOptions.speed * attractOptions.factor;
                const attractFactor = (0, NumberUtils_1.clamp)((0, NumberUtils_1.calcEasing)(1 - distance / attractRadius, attractOptions.easing) * velocity, 0, attractOptions.maxSpeed);
                const normVec = Vector_1.Vector.create(distance === 0 ? velocity : (dx / distance) * attractFactor, distance === 0 ? velocity : (dy / distance) * attractFactor);
                particle.position.subFrom(normVec);
            }
        }
        clickAttract() {
            const container = this.container;
            if (!container.attract) {
                container.attract = { particles: [] };
            }
            if (!container.attract.finish) {
                if (!container.attract.count) {
                    container.attract.count = 0;
                }
                container.attract.count++;
                if (container.attract.count === container.particles.count) {
                    container.attract.finish = true;
                }
            }
            if (container.attract.clicking) {
                const mousePos = container.interactivity.mouse.clickPosition;
                if (!mousePos) {
                    return;
                }
                const attractRadius = container.retina.attractModeDistance;
                this.processAttract(mousePos, attractRadius, new Circle_1.Circle(mousePos.x, mousePos.y, attractRadius));
            }
            else if (container.attract.clicking === false) {
                container.attract.particles = [];
            }
            return;
        }
    }
    exports.Attractor = Attractor;
});
