// tsparticles/umd/Plugins/Absorbers/index.js
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./Options/Classes/Absorber", "./Absorbers", "../../Utils/Utils", "./Enums/AbsorberClickMode", "./Options/Interfaces/IAbsorberOptions"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.loadAbsorbersPlugin = void 0;
    const Absorber_1 = require("./Options/Classes/Absorber");
    const Absorbers_1 = require("./Absorbers");
    const Utils_1 = require("../../Utils/Utils");
    class AbsorbersPlugin {
        constructor() {
            this.id = "absorbers";
        }
        getPlugin(container) {
            return new Absorbers_1.Absorbers(container);
        }
        needsPlugin(options) {
            var _a, _b, _c;
            if (options === undefined) {
                return false;
            }
            const absorbers = options.absorbers;
            if (absorbers instanceof Array) {
                return !!absorbers.length;
            }
            else if (absorbers) {
                return true;
            }
            else if (((_c = (_b = (_a = options.interactivity) === null || _a === void 0 ? void 0 : _a.events) === null || _b === void 0 ? void 0 : _b.onClick) === null || _c === void 0 ? void 0 : _c.mode) &&
                (0, Utils_1.isInArray)("absorber", options.interactivity.events.onClick.mode)) {
                return true;
            }
            return false;
        }
        loadOptions(options, source) {
            var _a, _b;
            if (!this.needsPlugin(options) && !this.needsPlugin(source)) {
                return;
            }
            const optionsCast = options;
            if (source === null || source === void 0 ? void 0 : source.absorbers) {
                if ((source === null || source === void 0 ? void 0 : source.absorbers) instanceof Array) {
                    optionsCast.absorbers = source === null || source === void 0 ? void 0 : source.absorbers.map((s) => {
                        const tmp = new Absorber_1.Absorber();
                        tmp.load(s);
                        return tmp;
                    });
                }
                else {
                    let absorberOptions = optionsCast.absorbers;
                    if ((absorberOptions === null || absorberOptions === void 0 ? void 0 : absorberOptions.load) === undefined) {
                        optionsCast.absorbers = absorberOptions = new Absorber_1.Absorber();
                    }
                    absorberOptions.load(source === null || source === void 0 ? void 0 : source.absorbers);
                }
            }
            const interactivityAbsorbers = (_b = (_a = source === null || source === void 0 ? void 0 : source.interactivity) === null || _a === void 0 ? void 0 : _a.modes) === null || _b === void 0 ? void 0 : _b.absorbers;
            if (interactivityAbsorbers) {
                if (interactivityAbsorbers instanceof Array) {
                    optionsCast.interactivity.modes.absorbers = interactivityAbsorbers.map((s) => {
                        const tmp = new Absorber_1.Absorber();
                        tmp.load(s);
                        return tmp;
                    });
                }
                else {
                    let absorberOptions = optionsCast.interactivity.modes.absorbers;
                    if ((absorberOptions === null || absorberOptions === void 0 ? void 0 : absorberOptions.load) === undefined) {
                        optionsCast.interactivity.modes.absorbers = absorberOptions = new Absorber_1.Absorber();
                    }
                    absorberOptions.load(interactivityAbsorbers);
                }
            }
        }
    }
    async function loadAbsorbersPlugin(engine) {
        const plugin = new AbsorbersPlugin();
        await engine.addPlugin(plugin);
    }
    exports.loadAbsorbersPlugin = loadAbsorbersPlugin;
    __exportStar(require("./Enums/AbsorberClickMode"), exports);
    __exportStar(require("./Options/Interfaces/IAbsorberOptions"), exports);
});
